### PR TITLE
V2.2.0

### PR DESCRIPTION
**Pull request by sasbury**
Nkey support.
Updated travis to 1.3 gnatsd - need to update again when we get nonce-auth released.
Made the build a bit more friendly to 'build from source' folks

### CODE REVIEW

File: src/main/java/io/nats/client/Options.java [hunk 1]
```diff
@@ -1234,7 +1270,14 @@ public String buildProtocolConnectOptionsString(String serverURI, boolean includ
...
 
-        if (includeAuth) {
+        if (includeAuth && nonce != null && this.getAuthHandler() != null) {
+            String nkey = this.getAuthHandler().getID();
+            byte[] sig = this.getAuthHandler().sign(nonce);
+            String encodedSig = Base64.getEncoder().encodeToString(sig);
+
+            appendOption(connectString, Options.OPTION_NKEY, nkey, true, true);
+            appendOption(connectString, Options.OPTION_SIG, encodedSig, true, true);
+        } else if (includeAuth) {
```
Comment (kozlovic) [inline src/main/java/io/nats/client/Options.java]: In Go client, when processing the info, if we get a nonce (that is, from a server that supports it), and the NKey option is not set (in your case, a auth handler) then we fail:
```
func (nc *Conn) processExpectedInfo() error {
(...)
	if nc.Opts.Nkey != "" && nc.info.Nonce == "" {
		return ErrNkeysNoSupported
	}
(...)
```
I am not seeing that in this file. Did I miss it?

File: src/test/java/io/nats/client/NKeyTests.java [hunk 2]
```diff
@@ -0,0 +1,525 @@
+// Copyright 2015-2018 The NATS Authors
```
Comment (kozlovic) [inline src/test/java/io/nats/client/NKeyTests.java]: Date in copyright

File: src/test/java/io/nats/client/TestAuthHandler.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2015-2018 The NATS Authors
```
Comment (kozlovic) [inline src/test/java/io/nats/client/TestAuthHandler.java]: Date

File: .travis.yml [hunk 4]
```diff
@@ -32,7 +32,7 @@ after_success:
 #Disable for now, upload archives fails because of IP address changes - "test ${TRAVIS_PULL_REQUEST} != 'true' && test ${TRAVIS_BRANCH} = 'master' && ./gradlew closeAndReleaseRepository"
 env:
   global:
-  - gnatsd_version=v1.2.0
+  - gnatsd_version=v1.3.0
   - gnatsd_path=$TRAVIS_BUILD_DIR/gnatsd/gnatsd
   - GPG_KEYRING_FILE=.travis/nats.travis.gpg
   - secure: yvOfk7kJzzTQ38n444jTDets24FZmxewwb3lrhXwpHTwOnQyq/B8QaHeqvhneECMc0Bq5M4blTlJ/wOWJAvs61POv2QVkyw+u8cVNROzkb8GPaH4ybPo8HMl33EHFNqh1KRo2C9hAPMYbbTjKCVY2UdkdfJ2l4lN/Awk7uEDX8ckc/sENhDeQjY/xoGZUP28O568Eg4ZxN3fr3WEV/0T+R15YyL2X0ev8MiGJM5TojXnNFKdb5fkUodRWwiY8JDn5xzP7xUzzen7MqE/5YNTcIC6haU8LToJM2gXEQtdoWLZqMPWr7k4A+eTBO5vl9qWrPBaOodFJYKzEjrEDfHj5RR9uaufEsnwQzXKw1ODrIFVZiC2n73j/tatWDI+vjnJ5tO+VMwWj53qdBYrvYeyewIT3cz9rrDHH8fGINsKAsk6HgWM3SMgeNSuXjRN0ePxEph5FVQ3ZUjF1ZXp90O7kjD5kXg/jVs6GrhCviRT3fx6Z4hyat9ytshy66jqcttHEfJ5sSOBg8fVbWJjLbxmghWUFp1fuc0HGNiMJStEyOBai5AkG6uJccTlgjlNL/8mgEF+fxo8HGVyStQzRnr7LJuCmWW9hx/aBVmqXR4p6cRgsSO09PvHRmcsLQoktCxVxsvcfblQqMbiQKjsJ4tXLe0U88DMOHnEGOgtik/tt+4=
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply .travis.yml]: Why do we have GPG key? Are we building gnatsd everytime? Can't we just download a release build?

File: src/main/java/io/nats/client/NKey.java [hunk 5]
```diff
@@ -0,0 +1,770 @@
...
+import net.i2p.crypto.eddsa.spec.EdDSANamedCurveTable;
+import net.i2p.crypto.eddsa.spec.EdDSAPrivateKeySpec;
+import net.i2p.crypto.eddsa.spec.EdDSAPublicKeySpec;
+
+class DecodedSeed {
+    int prefix;
+    byte[] bytes;
+}
+
+/**
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply src/main/java/io/nats/client/NKey.java]: +1, can always add it in if needed.

File: .travis.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sasbury) [reply .travis.yml]: maven

File: src/main/java/io/nats/client/NKey.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kozlovic) [reply src/main/java/io/nats/client/NKey.java]: @sasbury I would try to remove what can be removed or at least make the extra ones private. Once it is public, it more difficult to remove.

[the remark this answers is not in the document]
Comment (kozlovic) [reply src/main/java/io/nats/client/NKey.java]: By extra I meant anything that is not needed to sign the nonce. But I get your point about people not wanted to use something else to manage nkeys.

[the remark this answers is not in the document]
Comment (aricart) [reply src/main/java/io/nats/client/NKey.java]: I think the library should be complete, even on tests, you want to generate stuff. Sure there's a bunch of 'extra' from the perspective of the signing. In any case, the java library has the same functionality as the go library. So would this mean that the go library also needs to be paired down?

Another way to dice it is to have a nkey-signing library, that provides an API that is reduced, while still keeping all the functionality on the library as is.

[the remark this answers is not in the document]
Comment (kozlovic) [reply src/main/java/io/nats/client/NKey.java]: I don't see it that way. All the NKey features that we have in Go will be for tooling, and the good thing about being it in Go is that we could make simple executable for all platforms. Other lang that needs NKey support right now at least just need to sign.

[the remark this answers is not in the document]
Comment (aricart) [reply src/main/java/io/nats/client/NKey.java]: I see it as we have an nkey library that is used by a client but really could be used for other things as well.

[the remark this answers is not in the document]
Comment (derekcollison) [reply src/main/java/io/nats/client/NKey.java]: I'd prefer to add that if we get lots of people asking for it..

**Pull request merged**